### PR TITLE
sysvari: enable statsCache lru by config

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### What is changed and how it works?

1. add global variable tidb_stats_cache_mem_quota to control statsCache capacity
2. use lru cache for statsCache by config
3. make `initStats` / `handle.Update` / `AnalyzeTask` compatible with lru statsCache.

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/statscache.go
Comment (chrysan) [inline statistics/handle/statscache.go:43]: What about using variable.StatsCacheMemQuota directly when needed instead of passing domain->handle->statscache, since it's already a global atomic.

Comment (chrysan) [inline statistics/handle/statscache.go:43]: What about using variable.StatsCacheMemQuota directly when needed instead of passing domain->handle->statscache, since it's already a global atomic.

File: domain/sysvar_cache.go
Comment (morgo) [inline domain/sysvar_cache.go:262]: We just refactored this in https://github.com/pingcap/tidb/pull/33897

Unfortunately if you need the domain, you'll need to specially handle this. I don't think there's an easy way to access it from the SetGlobal func.

File: sessionctx/variable/sysvar.go
Comment (morgo) [inline sessionctx/variable/sysvar.go:685]: You don't need to specify skipInit. That has a special meaning for session scoped sysvars.

Comment (morgo) [reply sessionctx/variable/sysvar.go:685]: Also: is it possible to set the MaxValue to something like 1G? (or whatever makes sense as a ceiling that exceeds what is useful). We have had some problems in the past with misconfiguration, and try to keep max values to safe ranges.

File: statistics/handle/statscache.go
Comment (Yisaer) [reply statistics/handle/statscache.go]: updated.

File: domain/sysvar_cache.go
Comment (Yisaer) [reply domain/sysvar_cache.go]: updated.

File: sessionctx/variable/sysvar.go
Comment (Yisaer) [reply sessionctx/variable/sysvar.go]: However, I have no idea for a proper ceiling value for this variable for now.

File: statistics/handle/lru_cache.go
Comment (Yisaer) [inline statistics/handle/lru_cache.go:77]: Using copy here is necessary, otherwise it will cause data race like following:

```text
[2022-05-07T04:44:38.619Z] WARNING: DATA RACE
[2022-05-07T04:44:38.619Z] Write at 0x00c00a7b61a0 by goroutine 75:
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*statsInnerCache).put()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/lru_cache.go:117 +0xd1
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*statsInnerCache).Put()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/lru_cache.go:109 +0xc4
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.statsCache.update()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/statscache.go:91 +0x1e1
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*Handle).Update()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/handle.go:303 +0x32b
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/domain/domain.go:1390 +0x725
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker-fm()
[2022-05-07T04:44:38.619Z]       <autogenerated>:1 +0x39
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/wait_group_wrapper.go:33 +0x73
[2022-05-07T04:44:38.619Z] 
[2022-05-07T04:44:38.619Z] Previous read at 0x00c00a7b61a0 by goroutine 156:
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*statsInnerCache).Get()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/lru_cache.go:95 +0x10f
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*Handle).TableStatsFromStorage()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/handle.go:841 +0x167
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*Handle).Update()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/handle.go:285 +0x8cd
[2022-05-07T04:44:38.619Z]   github.com/pingcap/tidb/statistics/handle.(*Handle).LoadStatsFromJSON()
[2022-05-07T04:44:38.619Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/statistics/handle/dump.go:228 +0x3a8
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/executor.(*LoadStatsInfo).Update()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/load_stats.go:90 +0x25b
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*clientConn).handleLoadStats()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/conn.go:1756 +0xd7
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*clientConn).handleQuerySpecial()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/conn.go:2073 +0x34d
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*clientConn).handleStmt()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/conn.go:2045 +0x588
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*clientConn).handleQuery()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/conn.go:1875 +0x94f
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*clientConn).dispatch()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/conn.go:1370 +0x11f1
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*clientConn).Run()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/conn.go:1120 +0x2f3
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*Server).onConn()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/server.go:556 +0x1b84
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*Server).startNetworkListener.func2()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/server.go:453 +0x47
[2022-05-07T04:44:38.620Z] 
[2022-05-07T04:44:38.620Z] Goroutine 75 (running) created at:
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/wait_group_wrapper.go:31 +0xdc
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/domain/domain.go:1319 +0x27c
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/domain/domain.go:1299 +0x6b
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/session.BootstrapSession()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/session/session.go:2908 +0xac4
[2022-05-07T04:44:38.620Z]   main.createStoreAndDomain()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/tidb-server/main.go:298 +0x113
[2022-05-07T04:44:38.620Z]   main.main()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/tidb-server/main.go:204 +0x5e9
[2022-05-07T04:44:38.620Z] 
[2022-05-07T04:44:38.620Z] Goroutine 156 (running) created at:
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*Server).startNetworkListener()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/server.go:453 +0x6f3
[2022-05-07T04:44:38.620Z]   github.com/pingcap/tidb/server.(*Server).Run.func1()
[2022-05-07T04:44:38.620Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/server/server.go:368 +0x65
```

File: domain/sysvar_cache.go
Comment (xuyifangreeneyes) [inline domain/sysvar_cache.go:154]: Maybe we can do it in `SetGlobal`?

Comment (Yisaer) [reply domain/sysvar_cache.go:154]: It's hard to get `domain` in `SetGlobal`.

**Review by xuyifangreeneyes — APPROVED** [review]

File: statistics/handle/statscache.go
Comment (chrysan) [inline statistics/handle/statscache.go:43]: What about keeping mapCache so that we can fallback to mapCache if needed?

File: domain/sysvar_cache.go
Comment (chrysan) [reply domain/sysvar_cache.go:154]: Morgan just did a refactor to remove the specific logic here... And I do like this way:
https://github.com/pingcap/tidb/pull/33054/files#diff-93901abecfb2298268c7204c8665c636fa37285005c6b53fd3a552cd163c3e7eR887
Maybe we can have a discussion with Morgan to have an alignment.

File: sessionctx/variable/sysvar.go
Comment (chrysan) [inline sessionctx/variable/sysvar.go:681]: Why this MaxValue? If you want to define a meaningful maxValue, what about making it a const?

File: statistics/handle/handle.go
Comment (chrysan) [inline statistics/handle/handle.go:301]: tables are not put into oldCache yet, oldCache is not yet changed here.

**Review by chrysan — COMMENTED** [review]
I think the title should be changed to something like "enable lru for stats cache with quota", since it's not only adding new variable.
And it's better to have a config/variable to enable or disable lru cache, otherwise we have to make testings done and QA sign-off before we merge.

File: statistics/handle/statscache.go
Comment (Yisaer) [reply statistics/handle/statscache.go]: updated.

File: sessionctx/variable/sysvar.go
Comment (Yisaer) [reply sessionctx/variable/sysvar.go]: updated.

File: statistics/handle/handle.go
Comment (Yisaer) [reply statistics/handle/handle.go]: updated.

File: statistics/handle/lru_cache.go
Comment (Yisaer) [reply statistics/handle/lru_cache.go]: removed it now.

File: statistics/handle/lru_cache.go
Comment (Yisaer) [reply statistics/handle/lru_cache.go]: using Put instead now.

**Review by chrysan — APPROVED** [review]

File: domain/sysvar_cache.go
Comment (morgo) [reply domain/sysvar_cache.go:154]: Yeah, it's a known issue with getting the domain :( Unfortunately you have to do this trick in sysvar_cache.go.

File: domain/sysvar_cache.go
Comment (morgo) [inline domain/sysvar_cache.go:211]: @Yisaer can we change this to debug? I just rebuilt from master and its a little noisy. It re-runs whenever sysvar cache is rebuilt (every 30s).

**Pull request merged**
